02-KernelModules_CharacterDevice/chrdev.c

```diff
@@ -0,0 +1,232 @@
+	res = raw_copy_to_user((void*)buffer, msg - *offset, len);
+	*offset += len;
+	return len;
+}
+
+int dev_open(struct inode *inode, struct file *filp){	
+	if(busy){
+		printk(KERN_ERR "chrdev: failed to open");
+		return -EBUSY;
+	}
+	busy = true;
+	return 0;
+}
+
+int dev_release(struct inode *inode, struct file *filp){
+	busy = false;
+	return 0;
+}
+
+ssize_t dev_read(struct file *filep, char *buffer, size_t len, loff_t *offset){
+	ssize_t retval = 0;
+
+	int remain = buffer_size - (int) (*offset);
+	if (remain == 0) {
+		return retval;
+	}
+
+	if (len > remain) {
+		len = remain;
+	}
+
+	if (raw_copy_to_user(buffer, data_buff + *offset, len)) {
+		retval = -EFAULT;
+		return retval;
+	}
+	*offset += len;
+	retval = len;
+
+	return retval;
+}
+
+ssize_t dev_write(struct file *filep, const char *buffer, size_t len, loff_t *offset){
+	int error = 0;
+	int remain = buffer_size - (int) (*offset);
+	
+	used_memory += len;
+	
+	if(0 == remain) return 0;
+	if(used_memory > remain) return -EIO;
+	
+	error = raw_copy_from_user(data_buff + *offset, buffer, len);
+	if(error){
+		printk(KERN_INFO "chrdev: failed to receive %zu characters from the user\n", len);
+		return -EFAULT;
+	}
+	else{
+		printk(KERN_INFO "chrdev: received %zu characters from the user\n", len);
+		offset += len;
+		return len;
+	}
+}
+
+static struct file_operations fops = {
+	.open = dev_open,
+	.read = dev_read,
+	.write = dev_write,
+	.release = dev_release,
+};
+
+static int __init chrdev_init(void){
+	int err;
+	
+	printk(KERN_INFO "chrdev loading");
+	major = register_chrdev(0, DEVICE_NAME, &fops);
+	
+	major = register_chrdev(0, DEVICE_NAME, &fops);
+	if (major < 0) {
+		pr_err("register_chrdev failed: %d\n", major);
+		return major;
+	}
+	pr_info("chrdev: register_chrdev ok, major = %d\n", major);
+	
+	pclass = class_create(THIS_MODULE, CLASS_NAME);
+	if (IS_ERR(pclass)) {
+		unregister_chrdev(major, DEVICE_NAME);
+		pr_err("chrdev: class_create failed\n");
+		return PTR_ERR(pclass);
+	}
+	pr_info("chrdev: device class created successfully\n");
+
+    pdev = device_create(pclass, NULL, MKDEV(major, 0), NULL, DEVICE_NAME);
+	if (IS_ERR(pdev)) {
+		class_destroy(pclass);
+		unregister_chrdev(major, DEVICE_NAME);
+		pr_err("chrdev: device_create failed\n");
+		return PTR_ERR(pdev);
+	}
+	printk(KERN_INFO "chrdev: Device initialized succesfully\n");
+
+	if (0 == err) err = buffer_create();
+	if (0 == err) err = create_proc();
+	chd_kobject = kobject_create_and_add(MODULE_NAME, kernel_kobj);
+	if (0 == err) err = sysfs_create_file(chd_kobject, &sysfs_attribute.attr);
+	
+	return err;
+}
+
+static void __exit chrdev_exit(void){
+	kobject_put(chd_kobject);
+	
+	cleanup_proc();
+	buffer_clean();
+	device_destroy(pclass, MKDEV(major,0));
+	class_destroy(pclass);
+	class_unregister(pclass);
+	
+	unregister_chrdev(major, DEVICE_NAME);
+	printk(KERN_INFO "chrdev: Module unloaded\n");
+}
+
+module_init(chrdev_init);
+module_exit(chrdev_exit);
+
+MODULE_LICENSE("GPL");
```

Comment:
```suggestion
MODULE_LICENSE("GPL");
MODULE_AUTHOR("");
MODULE_DESCRIPTION("");
MODULE_VERSION("0.1");
```
Please fix description. Add  AUTHOR, version